File: arrays/07_two_sum/two_sum.cpp
#include <iostream>
#include <stdexcept>
#include <cmath>
#include <vector>
#include <unordered_map>
#include <cstdint>

using namespace std;


class Solution {
public:
    vector<uint16_t> twoSum(const vector<int>& nums, int& target) {
        if (!checkConstraints(nums, target)) {
            throw std::runtime_error("Constraints violated");
        }
        unordered_map<uint16_t, uint16_t> inds;
        for (uint16_t i = 0; i < nums.size(); i++) {
            int complement = target - nums[i];
            if (inds.find(complement) != inds.end()) {
                return {inds[complement], i};
            }
            inds[nums[i]] = i;
        }
        return {};
    }

private:
    // Helper function to check the constraints
    bool checkConstraints(const vector<int>& nums, int& t) {
        int n = nums.size();
        if (n < 2 || n > pow(10, 4)) return false;
        if (t < -pow(10, 9) || t > pow(10, 9)) return false;

        for (auto& num : nums) {
            if (num < -pow(10, 9) || num > pow(10, 9)) return false;
        }
        return true;
    }
};

File: arrays/07_two_sum/two_sum.py
from typing import List, Dict


class Solution:
    def count_paths(self, nums: List[int], target: int) -> List[int]:
        if not self._check_constraints(nums, target):
            raise ValueError('Constraints violated')

        inds: Dict[int, int] = {}
        for i in range(len(nums)):
            complement = target - nums[i]
            if complement in inds.keys():
                return [inds[complement], i]
            inds[nums[i]] = i

        return []

    def _check_constraints(self, nums: List[int], t: int) -> bool:
        n = len(nums)
        if n < 2 or n > 10**4:
            return False
        if t < -10**9 or t > 10**9:
            return False

        for num in nums:
            if num < -10**9 or num > 10**9:
                return False

        return True
